crates/bevy_render/src/render_resource/storage_buffer.rs

```diff
@@ -0,0 +1,138 @@
+    scratch: Vec<u8>,
+    storage_buffer: Option<Buffer>,
+}
+
+impl<T: AsStd430 + Default, U: AsStd430> Default for StorageBuffer<T, U> {
+    /// Creates a new [`StorageBuffer`]
+    ///
+    /// This does not immediately allocate system/video RAM buffers.
+    fn default() -> Self {
+        Self {
+            body: T::default(),
+            values: Vec::new(),
+            scratch: Vec::new(),
+            storage_buffer: None,
+        }
+    }
+}
+
+impl<T: AsStd430, U: AsStd430> StorageBuffer<T, U> {
+    // NOTE: AsStd430::std430_size_static() uses size_of internally but trait functions cannot be
+    // marked as const functions
+    const BODY_SIZE: usize = std::mem::size_of::<T>();
+    const ITEM_SIZE: usize = std::mem::size_of::<U>();
+
+    /// Gets the reference to the underlying buffer, if one has been allocated.
+    #[inline]
+    pub fn buffer(&self) -> Option<&Buffer> {
+        self.storage_buffer.as_ref()
+    }
+
+    #[inline]
+    pub fn binding(&self) -> Option<BindingResource> {
+        Some(BindingResource::Buffer(BufferBinding {
+            buffer: self.buffer()?,
+            offset: 0,
+            size: Some(NonZeroU64::new((self.size()) as u64).unwrap()),
+        }))
+    }
+
+    #[inline]
+    pub fn set_body(&mut self, body: T) {
+        self.body = body;
+    }
+
+    fn reserve_buffer(&mut self, device: &RenderDevice) -> bool {
+        let size = self.size();
+        if self.storage_buffer.is_none() || size > self.scratch.len() {
+            self.scratch.resize(size, 0);
+            self.storage_buffer = Some(device.create_buffer(&BufferDescriptor {
+                label: None,
+                size: size as wgpu::BufferAddress,
+                usage: BufferUsages::COPY_DST | BufferUsages::STORAGE,
+                mapped_at_creation: false,
+            }));
+            true
+        } else {
+            false
+        }
+    }
+
+    fn size(&self) -> usize {
+        let mut size = 0;
+        size += Self::BODY_SIZE;
+        if Self::ITEM_SIZE > 0 {
+            if size > 0 {
+                // Pad according to the array item type's alignment
+                size = (size + <U as AsStd430>::Output::ALIGNMENT - 1)
+                    & !(<U as AsStd430>::Output::ALIGNMENT - 1);
+            }
+            // Variable size arrays must have at least 1 element
+            size += Self::ITEM_SIZE * self.values.len().max(1);
+        }
+        size
+    }
+
+    pub fn write_buffer(&mut self, device: &RenderDevice, queue: &RenderQueue) {
+        self.reserve_buffer(device);
+        if let Some(storage_buffer) = &self.storage_buffer {
+            let range = 0..self.size();
+            let mut writer = std430::Writer::new(&mut self.scratch[range.clone()]);
+            let mut offset = 0;
+            // First write the struct body if there is one
+            if Self::BODY_SIZE > 0 {
+                if let Ok(new_offset) = writer.write(&self.body).map_err(|e| warn!("{:?}", e)) {
```

Comment:
is failing to write really just a warning? Should it bubble up the Result instead?

Reply:
I personally think this should be unwrapped (like we do in UniformVec). This will only fail if we haven't allocated enough space in the buffer. That is something we should be doing correctly (and behavior would be undefined if we didnt).  Not something we should be surfacing to the user, or allowing to continue if we did this wrong.